### PR TITLE
Candidate Validation/PVF: more fidelity of error metrics

### PR DESCRIPTION
**Pull request by mrcnski**
# PULL REQUEST

## Overview

Adds some more metrics buckets for the results of candidate validation (i.e. PVF preparation/execution). This was just supposed to be a fun one for me.

I had to add some more error variants along the way. The error being metric'd did not have enough fidelity (having just a single variant for the result of PVF prep/exec, that contained the original error but stringified).

I also split up `PrepareError`. We were handling it differently based on whether it is deterministic. Deterministic variants were already not allowed in certain places, and same for non-deterministic variants, so we might as well enforce this on the type level. This revealed an inconsistency with the way that timeout errors are treated (see TODO).

## Related Issues

Closes https://github.com/paritytech/polkadot/issues/3755

### CODE REVIEW

File: node/core/candidate-validation/Cargo.toml [hunk 1]
```diff
@@ -19,6 +19,7 @@ polkadot-node-primitives = { path = "../../primitives" }
 polkadot-node-subsystem = {path = "../../subsystem" }
 polkadot-node-subsystem-util = { path = "../../subsystem-util" }
 
+# TODO: Do we need this?
 [target.'cfg(not(any(target_os = "android", target_os = "unknown")))'.dependencies]
```
Comment (m-cat) [inline node/core/candidate-validation/Cargo.toml:23]: Why is this here? The `polkadot-node-core-pvf` dependency isn't gated like this anywhere else.

File: node/primitives/src/lib.rs [hunk 2]
```diff
@@ -227,12 +227,23 @@ pub type UncheckedSignedFullStatement = UncheckedSigned<Statement, CompactStatem
...
+	CompilationError(String),
+	/// Failed to prepare in some other way. This includes function panicking.
+	PreparationError(String),
+	/// Failed to execute `validate_block`. This includes function panicking.
 	ExecutionError(String),
 	/// Validation outputs check doesn't pass.
 	InvalidOutputs,
+	// TODO: Currently unused. Preparation timeouts are treated as non-deterministic, so this can
+	// never be instantiated, whereas execution timeouts are deterministic. Should this
+	// inconsistency be addressed?
```
Comment (m-cat) [inline node/primitives/src/lib.rs:242]: I noticed that preparation timeouts are treated as non-deterministic (potentially spurious, reported as `ValidationFailed`) whereas execution timeouts are considered deterministic (reported as `InvalidCandidate`). This might make sense if pre-checking is supposed to rule out bad PVFs, haven't looked into it TBH.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2194421

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2194544

File: node/core/pvf/src/error.rs [hunk 3]
```diff
@@ -79,16 +98,20 @@ impl fmt::Display for PrepareError {
...
 	/// The error was raised because the candidate is invalid.
 	InvalidCandidate(InvalidCandidate),
-	/// This error is raised due to inability to serve the request.
-	InternalError(String),
+	/// This error is raised due to inability to serve the request during preparation.
+	InternalPrepareError(NonDeterministicError),
+	/// This error is raised due to inability to serve the request during execution.
+	InternalExecuteError(String),
+	/// This error is raised due to inability to serve the request for some other reason.
+	InternalOtherError(String),
```
Comment (m-cat) [inline node/core/pvf/src/error.rs:106]: This ranks among the worst names I've ever come up with. 😬

File: node/core/candidate-validation/Cargo.toml (hunk 1 above)
Comment (bkchr) [reply node/core/candidate-validation/Cargo.toml:23]: Can probably be removed.

File: node/core/pvf/src/error.rs [hunk 4]
```diff
@@ -43,29 +81,10 @@ pub enum PrepareError {
...
 }
 
-impl PrepareError {
-	/// Returns whether this is a deterministic error, i.e. one that should trigger reliably. Those
-	/// errors depend on the PVF itself and the sc-executor/wasmtime logic.
-	///
-	/// Non-deterministic errors can happen spuriously. Typically, they occur due to resource
-	/// starvation, e.g. under heavy load or memory pressure. Those errors are typically transient
-	/// but may persist e.g. if the node is run by overwhelmingly underpowered machine.
-	pub fn is_deterministic(&self) -> bool {
```
Comment (bkchr) [inline node/core/pvf/src/error.rs:53]: Why did you remove this?

File: node/core/pvf/src/error.rs [hunk 5]
```diff
@@ -79,16 +98,20 @@ impl fmt::Display for PrepareError {
 pub enum ValidationError {
 	/// The error was raised because the candidate is invalid.
 	InvalidCandidate(InvalidCandidate),
-	/// This error is raised due to inability to serve the request.
-	InternalError(String),
+	/// This error is raised due to inability to serve the request during preparation.
+	InternalPrepareError(NonDeterministicError),
```
Comment (bkchr) [inline node/core/pvf/src/error.rs:102]: Why not forward the prepareerror directly?

The variants also don't need the postfix "Error".

File: node/core/candidate-validation/src/lib.rs [hunk 6]
```diff
@@ -560,18 +561,31 @@ async fn validate_candidate_exhaustive(
...
-			Ok(ValidationResult::Invalid(InvalidCandidate::ExecutionError(e))),
-		Err(ValidationError::InvalidCandidate(WasmInvalidCandidate::AmbiguousWorkerDeath)) =>
-			Ok(ValidationResult::Invalid(InvalidCandidate::ExecutionError(
-				"ambiguous worker death".to_string(),
-			))),
-		Err(ValidationError::InvalidCandidate(WasmInvalidCandidate::PrepareError(e))) =>
-			Ok(ValidationResult::Invalid(InvalidCandidate::ExecutionError(e))),
+		// Internal errors.
+		Err(ValidationError::InternalPrepareError(e)) =>
+			Err(ValidationFailed::Prepare(e.to_string())),
```
Comment (bkchr) [inline node/core/candidate-validation/src/lib.rs:566]: And why convert all errors to strings?

File: node/core/pvf/src/error.rs [hunk 7]
```diff
@@ -24,12 +24,50 @@ pub type PrepareResult = Result<Duration, PrepareError>;
...
+			Panic(err) => write!(f, "panic: {}", err),
+		}
+	}
+}
+
+/// Non-deterministic errors can happen spuriously. Typically, they occur due to resource
+/// starvation, e.g. under heavy load or memory pressure. Those errors are typically transient but
+/// may persist e.g. if the node is run by overwhelmingly underpowered machine.
+#[derive(Debug, Clone, Encode, Decode)]
+pub enum NonDeterministicError {
```
Comment (bkchr) [inline node/core/pvf/src/error.rs:70]: I get the naming of deterministic and non deterministic. However, I think this could be improved. I don't have a better terminology at hand right now, but I think it could be better.

File: node/core/pvf/src/error.rs (hunk 5 above)
Comment (bkchr) [reply node/core/pvf/src/error.rs:102]: And yeah, looking into the code that handles this error, we should name this variant `Prepare` and forward the `PrepareError` directly.

**Review by bkchr — COMMENTED** [review]
I mean I see what you want to do. However I'm not sure that this pr improves the current situation 🙈

File: node/core/pvf/src/error.rs (hunk 5 above)
Comment (m-cat) [reply node/core/pvf/src/error.rs:102]: These internal variants are only for non-deterministic errors, hence the `Internal` prefix and why I use `NonDeterministicError` here.

File: node/core/pvf/src/error.rs (hunk 4 above)
Comment (m-cat) [reply node/core/pvf/src/error.rs:53]: I figured that now we can just match on the `Deterministic` variant directly, so this function seemed extraneous.

**Comment by m-cat** [comment]
> I mean I see what you want to do. However I'm not sure that this pr improves the current situation 🙈

Why do you say that? I think we have a bit more fidelity and soundness of errors now.

File: node/core/pvf/src/error.rs (hunk 7 above)
Comment (m-cat) [reply node/core/pvf/src/error.rs:70]: Yeah, it's awkward. Maybe `InternalPrepareError` and `InvalidPrepareError`? IDK.

File: node/core/pvf/src/error.rs (hunk 4 above)
Comment (bkchr) [reply node/core/pvf/src/error.rs:53]: Yeah for sure, but calling this method is less code 🤣 But yeah, no strong argument here.

File: node/core/pvf/src/error.rs (hunk 5 above)
Comment (bkchr) [reply node/core/pvf/src/error.rs]: Ahh, I had read the From implementation not good enough! Sorry!

**Comment by bkchr** [comment]
> > I mean I see what you want to do. However I'm not sure that this pr improves the current situation 🙈
> 
> Why do you say that? I think we have a bit more fidelity and soundness of errors now.

Don't wanted to sound like it's bad or something! Maybe you could change ValidationError to only have two variants? InvalidCandidate and InternalFailure? Or Something like that?

My "general complain" is the huge number of different errors that you all handle differently in the upper layers. While it is probably only important if the candidate is invalid or we think that some machine error lead to the validation to fail?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2196570

File: node/core/pvf/src/executor_intf.rs [hunk 8]
```diff
@@ -40,8 +40,8 @@ use std::{
 // The data section for runtimes are typically rather small and can fit in a single digit number of
 // WASM pages, so let's say an extra 16 pages. Thus let's assume that 32 pages or 2 MiB are used for
 // these needs by default.
-const DEFAULT_HEAP_PAGES_ESTIMATE: u64 = 32;
-const EXTRA_HEAP_PAGES: u64 = 2048;
+const DEFAULT_HEAP_PAGES_ESTIMATE: u32 = 32;
+const EXTRA_HEAP_PAGES: u32 = 2048;
```
Comment (m-cat) [inline node/core/pvf/src/executor_intf.rs:44]: This random change was due to `clippy::pedantic` warning about casting a `u64` to a `usize` (down below).

**Comment by m-cat** [comment]
> Don't wanted to sound like it's bad or something! Maybe you could change ValidationError to only have two variants? InvalidCandidate and InternalFailure? Or Something like that?

It's all good! I'm open to feedback, I just don't see how this can be improved yet.

`ValidationError` had only two variants to start with, but I split it up so that I could forward a more precise error into the metrics buckets:

```rs
Err(ValidationError::InternalPrepare(e)) => Err(ValidationFailed::Prepare(e)),
Err(ValidationError::InternalExecute(e)) => Err(ValidationFailed::Execute(e)),
Err(ValidationError::InternalOther(e)) => Err(ValidationFailed::Other(e)),
```

```rs
Err(ValidationFailed::Prepare(_)) => &["internal failure (preparation)"],
Err(ValidationFailed::Execute(_)) => &["internal failure (execution)"],
Err(ValidationFailed::Other(_)) => &["internal failure (misc)"],
```

> My "general complain" is the huge number of different errors that you all handle differently in the upper layers. While it is probably only important if the candidate is invalid or we think that some machine error lead to the validation to fail?

Yeah, the error handling here does feel pretty messy, even before my changes. I wanted to get the better metrics fidelity without introducing even more error types. But I feel like we should have internal vs. invalid in the metrics, as well as prepare vs. execute, at a minimum.

**Comment by bkchr** [comment]
Do you really need that much detail on the metrics level? Isn't there valid/invalid/internal error enough? This should be enough to know when something is going on and for more detail you then need to look into the logs any way? Or do you see any advantage in having such a detail on the error variants in the metrics?

**Comment by m-cat** [comment]
> Do you really need that much detail on the metrics level? Isn't there valid/invalid/internal error enough? This should be enough to know when something is going on and for more detail you then need to look into the logs any way? Or do you see any advantage in having such a detail on the error variants in the metrics?

Yeah, good questions! I don't really know. I just assumed that having more buckets could be useful. Maybe someone else can weigh in on that.

**Comment by bkchr** [comment]
I can see that people for example would add some kind of alert when internal validation errors are going over a certain threshold, because that may means that your node is not working properly.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2197558

File: node/core/pvf/src/execute/queue.rs [hunk 9]
```diff
@@ -149,7 +149,7 @@ impl Queue {
 						break;
 					}
 				}
-				ev = self.mux.select_next_some() => handle_mux(&mut self, ev).await,
+				ev = self.mux.select_next_some() => handle_mux(&mut self, ev),
```
Comment (slumber) [inline node/core/pvf/src/execute/queue.rs:152]: Please avoid adding too many unrelated changes as it worsens review experiense as well as commit history after squash.

File: node/primitives/src/lib.rs [hunk 10]
```diff
@@ -255,6 +266,9 @@ pub enum InvalidCandidate {
 	CodeHashMismatch,
 	/// Validation has generated different candidate commitments.
 	CommitmentsHashMismatch,
+	/// The worker has died during validation of a candidate. See
+	/// [`InvalidCandidate::AmbiguousWorkerDeath`].
+	AmbiguousWorkerDeath,
 }
```
[the remark this answers is not in the document]
Comment (mrcnski) [reply node/primitives/src/lib.rs:272]: Hmm, interesting. Before I jump back into this though, do you think this PR is worth pursuing further? Are the metrics buckets useful to us?

[TBH I thought this would be an easy change, and wanted to do minimal refactoring just to get more metrics buckets (I guess I got a bit carried away with `PrepareError`). My point being that I don't really want to do more reworking of the errors. 😛]

File: node/core/pvf/src/error.rs [hunk 11]
```diff
@@ -24,12 +24,50 @@ pub type PrepareResult = Result<Duration, PrepareError>;
 /// An error that occurred during the prepare part of the PVF pipeline.
 #[derive(Debug, Clone, Encode, Decode)]
 pub enum PrepareError {
+	/// An error that should trigger reliably. See [`DeterministicError`].
+	Deterministic(DeterministicError),
+	/// An error that may happen spuriously. See [`NonDeterministicError`].
+	NonDeterministic(NonDeterministicError),
+}
```
[the remark this answers is not in the document]
Comment (mrcnski) [reply node/core/pvf/src/error.rs:31]: These variants do end up being useful actually, there's several places where only "deterministic" (or not) errors are allowed, and having `PrepareError` in such places didn't quite seem correct.

If we pursue this further, I would probably rename to `PrepareFailedError` and `PrepareInvalidError` or something (not totally sure, haven't looked at this in a while).

File: node/primitives/src/lib.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (bkchr) [reply node/primitives/src/lib.rs:272]: I think these are too many metrics. What should an operator do with all these metrics? They want to know if the node is running correctly and if not, they should be notified. So, `Valid`, `Invalid`, `Invalid_Your_Validator_Has_Issues` should be enough IMO, while the latest one would be the thing that operators use to add some kind of alert. When this metric starts rising quickly, something is probably wrong with their node. However, for a detailed analysis logs will be required.

[the remark this answers is not in the document]
Comment (slumber) [reply node/primitives/src/lib.rs:272]: I agree with Basti

[the remark this answers is not in the document]
Comment (mrcnski) [reply node/primitives/src/lib.rs:272]: Fair enough! I can close this PR, should I close the related issue as well?

[the remark this answers is not in the document]
Comment (sandreim) [reply node/primitives/src/lib.rs:272]: +1 metrics should provide a birds eye view for edge cases or things not working correctly in general. Sometimes having more details (more metrics) makes sense but I've only seen that useful for performance tracking.
